### ISSUE TITLE
Android binary

### ISSUE THREAD

**Issue by acristescu** [lightvector/KataGo#321]
Now that we have a CPU-only version of the engine, would it be possible to add arm7 and/or arm64 to the compiled binary of the release? I couldn't get it going, but then I haven't compiled any C++ since I finished uni many years ago :) It should be possible though as there are several of these floating around for both Leela Zero and SAI (for example https://github.com/Grant-Tao/compiled-leelaz-0.17-for-android-phones and https://github.com/evdwerf/leela-zero/tree/android ).

The second repo above even has the modifications to the Makefile for Android (see this comparison https://github.com/leela-zero/leela-zero/compare/next...evdwerf:android ).

Is that something that could be achieved?

**Comment by simonguoxm** [comment]
this will be a great feature for the phone.

**Comment by l1t1** [comment]
ahq weiqi pro includes it.

http://www.flygo.net/bbs/forum.php?mod=viewthread&tid=112497&extra=page%3D1&mobile=2

**Comment by cryptpark** [comment]
I seem to have the wrong site. do you have another link? is this program paid?

**Comment by portkata** [comment]
acristescu will probably have one in a few months, until then we have https://github.com/aki65/aki65.github.io/releases
version .1 and .2 are opencl for qualcomm phones, .4 and .5 are eigen cpu for any phone, tablet or chromebook. Speed is good but i wish the katrain calibrated ai equation could be applied and distributed as a paid app. Everyone is doing great work with katago, it is a dream come true

**Comment by acristescu** [comment]
@portkata could you please describe what do you mean by the "calibrated ai equation"? FYI after I posted this, somebody on reddit pointed the aki65 stuff to me and I manage to reverse engineer the binary out of it. It's still not great as the guy did not put the scripts online, just the resulting APK and he or she might not build a newer version. Not having an official build is a big miss IMHO.

I am planning to include the AI into the free app OnlineGo (https://play.google.com/store/apps/details?id=io.zenandroid.onlinego&hl=en). I have solved most technical problems I think, I just need to tweak some things so that it doesn't beat us mere mortals too badly :)

**Comment by portkata** [comment]
@acristescu  Thank you for your work :)
Sanderbaduk uses an equation to weaken katago on one visit to play at various ranks 15k to low dan, but weaker than the policy. It is explained in  this issue thread in his katrain software https://github.com/sanderland/katrain/issues/74 and also referred to in passing here https://github.com/sanderland/katrain/issues/73
Incorporating this into a katago android gui if possible would let users of any rank play an even match against katago. It gives the user the ability to set katago's rank to weaker than 3d (1 visit). Your gui is the best android go experience so far, allowing the user to play against katago at >1500 visits/move, but it is not possible to be competitive againstnsuch a strong bot. It is still such great joy to play.

**Comment by portkata** [comment]
> ahq weiqi pro includes it.
> 
> http://www.flygo.net/bbs/forum.php?mod=viewthread&tid=112497&extra=page%3D1&mobile=2
link to download http://aqgo.ezandroid.cn/AQGoPro_1.3.0_2020-09-27.apk
Paypal ($15 US): 397093109@qq.com
activation code arrives in one business day.

**Comment by acristescu** [comment]
What?!?! $15 for just integrating the hard work of other people into an android app? That is highway robbery... Now I feel like a schmuck for giving the app for free :)

As for Katrain, yes, I am aware of that (see for example this reddit [thread ](https://www.reddit.com/r/baduk/comments/hejz5h/katrain_new_features_and_easy_katago_latest/)in which both me and @sanderland are discussing it). Thank you for pointing out the equation in the Katrain code, I'm going to give it a honest go at implementing it in the app, but @sanderland put a lot of work and thought into this I'm pretty sure I will not be able to do nearly as good a job as him/her. Perhaps @sanderland could be persuaded to contribute that aspect to my open-source app once I manage to get the integration going? :)

**Comment by sanderland** [comment]
@acristescu wait until you see things like [this](https://www.ebay.com/itm/HP-z220-i5-8-250GB-Win10-AI-Go-Weiqi-Baduk-Leela-ELF-mini-Go-Gaming-GTX1070-/283985680529) - when you pick a license you know what people can do with it, and I can't get worked up about it.
AI code is [here](https://github.com/sanderland/katrain/blob/master/katrain/core/ai.py) and is quite simple overall. KaTrain being based on kivy should be able to run on android, but no luck so far from my end.

**Comment by portkata** [comment]
@sanderland would it be very difficult for you to write the code for the calibrated ai in @acristescu's apk if he does the gui and katago implementation? Thanks for the time.

**Comment by acristescu** [comment]
That would be awesome, otherwise I'll have to "git gud" at magic such as this:

```    elt = [(math.log(random.random()) / (item[1] + 1e-18), item) for item in items]  # magic
    return [e[1] for e in heapq.nlargest(pick_n, elt)]  # NB fine if too small```

**Comment by sanderland** [comment]
well if you ignore the name of the function, which is `weighted_selection_without_replacement`....
I'm hesitant to get pulled into other projects as my own is already taking far more than I ever imagined. Please try it yourself first, I bet the main work is mapping things onto your own data structures, which I wouldn't know about anyway.

**Comment by portkata** [comment]
Maybe the calibrated ai was too much to ask. Does this zip file have all the fils necessary to run katago in android in the command line? https://drive.google.com/file/d/12QNGdFx0e0GHfpzB6GFWcsHoq1TRJklg/view?usp=drivesdk

**Comment by acristescu** [comment]
@portkata all that you need is available in the APKs you've initially posted ( https://github.com/aki65/aki65.github.io/releases ). You just have to unzip the APK, then locate a file called `private.mp3` then rename that as `.tar.gz` then unzip that, then find the katago executable. What you need are the binary (which is the one that this bug asks for as an official release) + any net + an appropriate cfg file. I have all that and I am not migrating my app from using GTP to using the katago JSON analysis engine (which IMHO is a lot nicer). I should have that done shortly, and then I can release a version without scaling. I expect that to play at 3-4d. Once that is done, I can, in a second iteration, try to reverse engineer @sanderland 's work (for which I am grateful, please don't misinterpret my "magic" comment as dismissive).

That being said, I still think the initial request of this issue is a valid one: a katago binary that I don't have to basically steal from somebody that is ripping people off with $15 price tags would be great (but beyond my skill).

**Comment by portkata** [comment]
Thanks for all the detailed replies and hard work towards this. Sorry, probably stupid, but just one last question. So if I'm understanding correctly the 3 files in this zip (android binary, 10b net, and gtp.cfg 14mb) https://drive.google.com/file/d/12XbVx4ebKikzxAhxyiFPtF0rIPpHr6o4/view?usp=drivesdk are the only ones needed for katago on 64 bit android command line. no dll file equivalents are needed? Thanks so much!

**Comment by acristescu** [comment]
I'm not sure where that zip file is coming from, but the binary in the aki65 repository above does not need any `.so` files (the equivalent of DLL). It just worked on my S9+ with Android Q.

**Comment by portkata** [comment]
> That being said, I still think the initial request of this issue is a valid one: a katago binary that I don't have to basically steal from somebody that is ripping people off with $15 price tags would be great (but beyond my skill).

Maybe @lightvector can ask akigo how he compiled the android binary? Akigo is fluent in English. Then if lightvector puts it on his github page, the community can update it in the future if the main branch speed is optimized further in later releases. It was a great idea to bring this up

**Comment by aki65** [comment]
From issue #74 in KaTrain it seemed to me, that the calibration was recently adapted to 20b, but I may be wrong. Surely, @sanderland can clarify here.

And no, I did not change anything from KaTrain's algorithm. I just allowed entering up to 9-dan for experimental purposes to see how the formulas behave when the rank is extended beyond 4-dan. Should I cut this off ?

**Comment by acristescu** [comment]
Hi @aki65 and welcome to the party. I think the issue here is not with Sanderland's formula (although we got a bit side-tracked here), the request was to basically integrate whatever you did to create an android binary into the main release so that there is an official android binary. Is that something that can be achieved?

**Comment by portkata** [comment]
I had the impression it was based on the 20b too, but he said 15b when asked a couple days ago on reddit, so not 100% sure. Either will give you good results. The policy network at full strength is roughly 4 dan or so, so higher numbers, won't do much. A better thing would be to set all the integers above his highest "kyu-rank" value to not use the algorithm but instead run katago using only 1 thread and x number for "maxPlayouts" in the gtp instead of maxvisits. If it was me, I'd set it based on the 20b (which does work well with his algorithm as well) and do 6 playouts for the 4 dan equivalent q rank, 8 for 5 dan, 10 for 6 dan, 12 for 7 dan, 14 for 8 dan, and 16 for 9 dan. Thanks to everyone for all these great projects, it's so exciting!

**Comment by sanderland** [comment]
> From issue #74 in KaTrain it seemed to me, that the calibration was recently adapted to 20b, but I may be wrong. Surely, @sanderland can clarify here.

it's still on 15b since that's the default, but the strengths are approximate anyway and they don't differ that much between 15b and 20b due to the weirdness of the algorithm.
 
> And no, I did not change anything from KaTrain's algorithm. I just allowed entering up to 9-dan for experimental purposes to see how the formulas behave when the rank is extended beyond 4-dan. Should I cut this off ?

It doesn't make much sense, as it converges to pure policy-based play for strength to infinity, which is around ogs 5 dan. At these levels the lack of reading becomes clearer as well, and players get a bit annoyed at how the AI seems to die.

**Comment by acristescu** [comment]
Since we're on the subject, which net should I bake into the app? So far I've been using the 10b one (g170e-b10c128-s1141046784-d204142634.bin) and it's reasonably fast, however it seems to me weaker than LeelaZero (15b-270 0c4ade) running on the same phone. It's kinda hard to say since I am about 10k and both of them beat me to a pulp, but Kata does make some weird moves such as putting both first move in the same corner instead of grabbing an empty one.

So the bottom line is, which net should I choose, keeping in mind that APK size is a concern (these nets are bigger than my entire APK!!!)?

**Comment by sanderland** [comment]
The 6b and 10b are really stepping stones in early phase training, the 15b went a lot longer, and the 20b/30b/40b to the end.
I use 15b by default for size concerns, it's superhuman, reasonably fast and decent at play and analysis. When analysing on desktop I use 20b myself, and I've barely touched the other nets.

**Comment by acristescu** [comment]
Thank you, b15 it is...

Is there a way to force kata to accept a model file that is not ending with .gz? The way Android resources are packed in an APK destroys archives, and kata is just throwing a fit if I rename the file:

```terminating with uncaught exception of type StringError: Error loading or parsing model file katago.net.so: Model file should end with .txt, .bin, .txt.gz, .bin.gz, or possibly just .gz. (If it doesn't
 have one of these extensions already, it's probably the wrong file, renaming will probably NOT help).```

**Comment by portkata** [comment]
would  it be possible for you to be able to package the .gz net file as contained in a folder inside of an archive with an extension you change to .mp3, then have the .mp3 file extension be changed and unpacked upon installation?

**Comment by acristescu** [comment]
I don't have write permissions in that folder so I can't rename it (crazy Android 10 restrictions). I can copy them manually to another temp folder under the right name if they don't already exist... That's the workaround I'm doing right now, but it's a hassle, I was hoping there's a `-model-i-know-what-im-doing-just-take-the-bloody-file` parameter :)

**Comment by sanderland** [comment]
@acristescu : there is not, since it needs the extension to figure out what to do with the file. See desc.cpp
Easy enough to edit in an extra extension if you're compiling it though.

**Comment by lightvector** [comment]
@acristescu 
https://github.com/lightvector/KataGo/blob/master/cpp/neuralnet/desc.cpp#L1111

Take a look at the code starting at that line and below. KataGo uses the extension to figure out whether it should try to load the file as compressed or uncompressed, and what format it is, because there are multiple possible formats that it tries to accept, for legacy reasons. As Sander said, if you're already compiling it yourself, a quick workaround would just be to edit the code here a little to just make it so that the extension that you want to use falls into the proper case instead of hitting the `else` at the very end that reports the failure.

**Comment by aki65** [comment]
@acristescu 

> ... the request was to basically integrate whatever you did to create an android binary into the main release so that there is an official android binary. Is that something that can be achieved?

I think so, but there is still work to be done. Keep in mind that my goal was just to create a working binary within my personal development environment in order to get BadukAI up and running. Therefore I had no qualms to spread machine-dependent settings and paths all over my scripts. This has to be factored out as far as possible before sharing them makes any sense. Furthermore, I just made every code syntax change in KataGo that was necessary to make the android cross-compiler happy. Some of these changes will probably break builds for other platforms. So I have to refine this (probably using some #ifdef directives) before the code changes can be merged into the KataGo repository.
All this is on my TODO list, but I am very busy at the moment with my day job and feature requests for BadukAI, so it will take some time ...

**Comment by acristescu** [comment]
Is there a new version of the android binary please? I spent way too much time figuring out that the parameter `includeMovesOwnership` that I'm trying to use has only been added 24 days ago, meaning after @lightvector compiled his one-time binary...

**Comment by acristescu** [comment]
On second thought, the release is new, but the katago binary seems to be the same (old) version (7,942,024 bytes). Is there a way to  get a newer version with all the new stuff by any chance?

**Comment by portkata** [comment]
wow that looks so good.

**Comment by acristescu** [comment]
I've released a first version of it: https://play.google.com/store/apps/details?id=io.zenandroid.onlinego using aki65's binary.

For now it has not scaling down functionality, it just plays at 30 visits with the 15b network which makes it WAY above my level. I will try my hand at dumbing it down, as it seems like a missed opportunity given the fact that the app knows your OGS rank and could thus scale katago down to your level automatically.

Thank you guys for your help.

PS: it would still be great to have an official android build with the latest and greatest though

**Comment by portkata** [comment]
The ui looks great but I can't figure out how to play katago online. I'm wondering if the new apk is not being hosted by google play?

**Comment by acristescu** [comment]
285 doesn't have it, 301 and 302 do. Sometimes play store is slow to suggest updates, but for me it did update. Try it again later pls.

**Comment by lightvector** [comment]
I don't build android (had someone else been doing this?) But you are right that the latest official release is lagging a bit behind master branch in features. I'll make a new release (for windows+linux) before too long. Thanks for the reminder.

**Comment by acristescu** [comment]
Hi @lightvector thank you for answering. Indeed, an android official build was the initial ask of this issue before it got completely derailed. In summary, yes, somebody has done it... as a one-time thing: @aki65 but as he explains [here](https://github.com/lightvector/KataGo/issues/321#issuecomment-701705506) he didn't get the build scripts in a good enough shape to raise a PR or even to make public.

Anyway, this is the binary I managed to bake into my app, however it is a bit out of date already (and it will get more so as time passes). I can confirm that his build does work quite nicely though and the performance is adequate for casual use. I feel it would be a great addition to the project to add android as an officially supported platform.

**Comment by sanderland** [comment]
@acristescu @portkata what kind of visits are you getting? I'm getting around 1/s on my phone and 0.5/s on my tablet on 15b.

**Comment by acristescu** [comment]
My S9+ gets about 7 visits/s. Other users of my app are reporting around 10 visits/s on S10 and 4-5 visits/s on A51. If you don't mind me asking, what phone and tabled are you getting those results on?

**Comment by sanderland** [comment]
S10 for the 1 visit/sec. Specifically using the analysis engine with an 8s timeout I get ~10 visits.

**Comment by acristescu** [comment]
Hmm... weird... I am also using the analysis engine (excellent work btw, that's much nicer than GTP), but with `maxVisits = 20`. What app are you using? Are you getting better speeds from mine?

**Comment by sanderland** [comment]
Your app does move in about 2 seconds. Maybe my settings are off. Is your numSearchThreads high?

My contributions to the analysis engine are quite small, by the way, but I agree it's a nice thing to have. :)

**Comment by portkata** [comment]
Wow i can't believe you were able to bring katrain to android, what a great thing :) . I get less than 2 visits/second with 1 thread 64 bit snapdragon 625. That is good enough for me I think.

**Comment by acristescu** [comment]
Wow, katrain on Android... that's going to be stiff competition :) Anyway, here's the CFG I use: https://github.com/acristescu/OnlineGo/blob/master/app/src/main/assets/katago.cfg

I haven't actually done any tests for speed, as for playing against katago visits / s are not important, it will beat 99% of the users in a couple of seconds anyway.

Btw, katrain was python, right? maybe it's the python platform that slowing things down?

**Comment by sanderland** [comment]
i suspect it's `numSearchThreadsPerAnalysisThread = 8`, the opencl is irrelevant right as it's eigen?
katrain is python, and python-for-android is a bit slow to start, but after that no amount of button drawing is going to compete with neural network evaluations in terms of turning your phone into a handwarmer.

**Comment by acristescu** [comment]
> i suspect it's `numSearchThreadsPerAnalysisThread = 8`, the opencl is irrelevant right as it's eigen?

I have absolutely no idea, I figured my phone has 8 cores, it can't hurt, right? If you do experiment and find some interesting settings for the CFG, please do share though...

> neural network evaluations in terms of turning your phone into a handwarmer.

Leela Zero did warm the phone right up, katago doesn't so much. Whatever magic aki65 did to create that build worked like a charm...

**Comment by isty2e** [comment]
You might be interested in using [ONNXRuntime backend](https://github.com/lightvector/KataGo/pull/341), which [supports NNAPI](https://github.com/microsoft/onnxruntime/blob/master/docs/execution_providers/NNAPI-ExecutionProvider.md).

**Comment by sanderland** [comment]
are there binaries for 1.7.0 out yet?

**Comment by acristescu** [comment]
> are there binaries for 1.7.0 out yet?

Not that I know of. No `includeMovesOwnership` for us... :(

**Comment by portkata** [comment]
aki65 optimized the newest version of his app so when it runs with his optimized 20b net, the net is 5 times faster :)

**Comment by acristescu** [comment]
So he modified the 20b network to work faster on Android 64-bits? Is there any loss of the quality of play? Which net did he pick to optimize (there are quite a few 20b ones).

**Comment by portkata** [comment]
@acristescu I think there's no loss in quality of play. Its probably the last 20b extended training, i can check it with a hex editor next weekend. With 1 thread on my device, it does 12 playouts in the same time it used to do 2!

**Comment by cryptpark** [comment]
network remains the same, its use is optimized

**Comment by acristescu** [comment]
Not quite true, what it looks like he did is converted the model to Google's TensorFlow lite format and somehow made KataGo work with tensorflow. This is quite impressive, as the new format should be both faster (and importantly for mobile) quite compact as well. I'm not sure if @lightvector knows about this, but I think this would be something really cool to include back in the main project.

However, I could not get it to run, probably because the command line parameters are different. @aki65 could you kindly share what the new command lines parameters are (in the spirit of open-source)?

**Comment by cryptsport** [comment]
unclear. if unpack the apk, then unpack "private.mp3", like a zip, there is a part of the code that completely coincides with the official network. (7ED1600h arm64-v8a-rel-0.21)

**Comment by acristescu** [comment]
The old network is still there, but now there's a new file called `20b.tflite`. If you try to just run the katago binary included in 0.21 you also get an error saying that it's missing `libtensorflowlite.so`. This did not happen with the old 0.16. From this I have deduced that he somehow made katago work with TensorFlow.

**Comment by cryptsport** [comment]
this is very interesting, but it doesn't work for me v64. so is there a new network or not? or is it  not clear?

**Comment by acristescu** [comment]
only @aki65 can answer these questions for sure.



> it doesn't work for me v64

Do you mean LazyBaduk does not work for you or that you cannot run the new katago binary? I can't get it to work either...

**Comment by cryptsport** [comment]
only 32bit version works on my devices. for some reason also on the nox emulator (probably there is no 64bit support, I will find out)

**Comment by cryptsport** [comment]
do you mean "BadukAI"? and which network? Can this network be pulled out of the apk? how to find?

**Comment by cryptsport** [comment]
the sizes of unpacked files "private" v17 and v18 - 198MB and 225MB. little space for another network

**Comment by portkata** [comment]
imagine after distributed training when the 40b policy reaches 9 dan and he optimizes that network. The kyu rank mode will be awesome. with the opening book on, they play well all the way down to kyu rank 10. they go up to kyu rank -8.

**Comment by acristescu** [comment]
All of that is great, too bad is not open source so that other open source projects could use it... :|

**Comment by cryptsport** [comment]
answer @aki65
https://github.com/aki65/aki65.github.io/issues/8#issuecomment-740121624

**Comment by l1t1** [comment]
what are the meaning of these files in the apk
```
10b.bin.gz
20b.bin.gz
15b.tflite
20b.tflite
40b.tflite
```

**Comment by portkata** [comment]
the 15b and 40b are for leela zero, they 20b is for the optimized katago, which is now faster than leela zero, setting search threads to 1.

**Comment by portkata** [comment]
> answer @aki65
> [aki65/aki65.github.io#8 (comment)](https://github.com/aki65/aki65.github.io/issues/8#issuecomment-740121624)

to summarize for this thread, aki65 said there should be no change in invoking katago to use the tensorflow optimized net. Can someone run the optimized net in android command line and confirm? This net is really great for calibrated ai because it is so much faster. One can run the kyu rank bots on very old mobile processors now with instant moves; together with his opening book for the first moves, the kyu rank bot plays great and fast from kyu rank 10 and stronger.

**Comment by acristescu** [comment]
> Can someone run the optimized net in android command line and confirm?

I just tried by unzipping the files provided, using `adb push` to upload them to an S9+, then doing `adb shell` and running the binary with `LD_LIBRARY_PATH=. ./katago_binary_android`. When providing no parameter it does work (prints the usage info), but as soon as I add any parameter (for example `LD_LIBRARY_PATH=. ./katago_binary_android version`) it stops working. It just exits without any error.

I have spent some 3 hours during the weekend trying every which way, with different versions of the libraries command lines, etc. The same method worked for the old katago, leela zero and SAI. I must be missing something...

**Comment by portkata** [comment]
just saw the new icon in your app, it looks great! I still wish katago used half as many playouts though. Thanks for putting in all the time!

**Comment by acristescu** [comment]
Plan A was to put in a scaling mechanism as the one in katrain, but I just can't find the time for such an undertaking.

Plan B was to basically have the app compute the number of playouts per second it gets and then scale the number of playouts for the next moves so as to keep the time per move to 1s on any device. This requires less development, but would make the AI be inconsistent.

Plan C would be to at least temporarily have a global setting in the settings page where you can tweak this to your heart's desire. Might go with this in the next version.

**Comment by portkata** [comment]
oh wow this is going to be great, thanks!

**Comment by portkata** [comment]
aki65 added support for the new distributed kata1 weights on android https://github.com/aki65/aki65.github.io/releases

**Comment by l1t1** [comment]
> aki65 added support for the new distributed kata1 weights on android. https://github.com/aki65/aki65.github.io/releases

thanks

**Comment by portkata** [comment]
aki65 released optimized s580 40b distributed training weight. It is extremely, extremely fast. https://github.com/aki65/aki65.github.io/releases/tag/v1.4.1

**Comment by acristescu** [comment]
If we could also have that open-source, that would be a dream...

**Comment by HackYardo** [comment]
Sorry to bother you all after so long! I'm an amateur in programming and **I'm not sure whether this Issue is discussing a full apk or something that just works?** Because I can't stand LeelaZero only can play Go games on a 19x19 board, which is too big on a phone screen. Today, I try to compile KataGo in Ubuntu in Termux on Android, it works and works fine, **all you need to do is to read the guide here**:
https://github.com/lightvector/KataGo/blob/master/Compiling.md#linux

**Comment by jopdorp** [comment]
I was able to compile for android with opencl support using my branch:
https://github.com/jopdorp/KataGo/tree/android-support

now in the runtime when I try to use it, I get a crash during the tuning process:


No existing tuning parameters found or parseable or valid at: /data/user/0/nl.jopdorp.opengoban/files/.katago/opencltuning/tune11_gpuMaliG710r0p0_x19_y19_c128_mv8.txt
Performing autotuning

......

Tuning hGemmWmma for convolutions
error: couldn't allocate output register for constraint 'r'

**Comment by jopdorp** [comment]
@acristescu do you use OpenCL in the Sente app? If so how did you do the tuning, and would you have some tune files I could try?

**Comment by Entract** [comment]
I guess this has been closed for a while.
I'm trying to get either a libkatago.so or libleelaz.so with JNI functions.
I see John Yu had a program for this, but its about 7 years old, and updating has proven tricky (for me).
https://github.com/John-Yu/leelaApplication/  << This would be pretty much exactly what I'm looking for but up to date and for arm64-v8a architecture
